### PR TITLE
NIP-101 : Event Compression and Encoding Support

### PR DESCRIPTION
**Pull request by b35363**
_NIP-101 Draft Proposal_
_Update on NIP-79 specifications_ 

### NIP-101 : Event Compression and Encoding Support
This NIP proposes a modification to the Nostr protocol to introduce a more flexible and modular approach for message compression and encoding. It suggests extending the current WebSocket message structure to include optional encoding information, such as `base64`, `base58`, `gzip64`, `gzip58`, and `plaintext`. This enhancement aims to improve modularity, retro-compatibility, and the potential for future encoding schemes.

### NIP-79 : `window.nostr` offline message signature & verificiation

Authentication standards like NIP-07 and NIP-47 have proven effective. However, there's a need to extend the capability of message signature and verification to third-party applications that don't utilize the NIP-01 standard. This is crucial for scenarios where users and remote entities require mutual verification without relying on the NIP-42 standard, which involves submitting events for network-based verification.

_I updated the NIP-79 standard proposal to clarify the proposition, I'd love to know you thaughts on it_

### CODE REVIEW

**Review by b35363 — COMMENTED** [review]
Depending on how relays implement message parsing, we might need to change some assertions... the root message size for `EVENT` is now variable between 2 and 3 of length.

It's great for backward comp. but I don't know for shure how old relays might respond to a 3 lengthed message.

**Comment by b35363** [comment]
#### Here are some benchmarks between plain JSON and different CODEC used
The proposal focuses on JSON vs gzip64 but we can easilly imagine encodings based on many more standards.

image
image
image

**Comment by Egge7** [comment]
If this is to be implemented, the. It should specify MUST behaviour in the case that no CAP message is exchanged. That way backwards compatibility is ensured.

**Comment by vitorpamplona** [comment]
I would split this into 2 PRs. NIP-79 feels like it could be a simple addition of sign and verify methods to NIP-07 and NIP-46. That could be done very easily. It doesn't even need to be a new NIP. You can just change existing NIPs to add the two methods. I don't even think you need to explain why they are needed. 

NIP-101 needs some testing and evidence gathering. We need to see which relays already offer compression and how well that compression is doing. Then compare to how well a new algorithm could do in the current proposal. The delta between the two is the gain that must be compared against the added complexity costs if Nostr adopts it.

**Comment by b35363** [comment]
> If this is to be implemented, the NIP should specify MUST behaviour in the case that no CAP message is exchanged. That way backwards compatibility is ensured.

Thanks! Well as described, the default encoding suggested is plaintext wich is itself optional, by default, without proper CAP exchange initiated by the server, the client must go plaintext, no need to specify it then.

I'd love to know how you had seen this.

**Comment by b35363** [comment]
> I would split this into 2 PRs. NIP-79 feels like it could be a simple addition of sign and verify methods to NIP-07 and NIP-46. That could be done very easily. It doesn't even need to be a new NIP. You can just change existing NIPs to add the two methods. I don't even think you need to explain why they are needed.
> 
> NIP-101 needs some testing and evidence gathering. We need to see which relays already offer compression and how well that compression is doing. Then compare to how well a new algorithm could do in the current proposal. The delta between the two is the gain that must be compared against the added complexity costs if Nostr adopts it.

Yep, I do totally agree, nip-79 does not even need a nip, I got scholed by some frustrated people on this, glad some are on the same page.

I'll propose nip 07 modifications

For nip-101 I made some benchmarks between json plaintext and different encodings, the overall conclusion is that in some cases where storage efficiency abd network cap are essential, compression codecs are bangers.

Not for all cases obviously, sometimes, where theres not a lot of events, nor big enough events, or we have strong computing limitations, it can be overkill.

I concluded that by letting the choice, we could take the best of both worlds.

**Comment by fiatjaf** [comment]
I don't get this, sounds like a terrible idea that we shouldn't do at all. You're compressing the JSON then base64ing it? In any case the same arguments I have against having a binary format and everything else also apply to this: it's not worth it.

**Comment by mattn** [comment]
CAP seems to be good idea for generic way to do something, I think. But I prefer that the EVENT message is more structured data format.

* 4-bytes length of data network byte ordered.
* encoded data with the agreed encodings between client and relay.

File: 101.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+> Note that when not provided, the default encoding is set to default, for backward compatibility.
+
+#### Supported Encodings
+
+This NIP proposes the inclusion of the following encoding identifiers:
+
+- `base64`: Base64 encoding.
+- `base58`: Base58 encoding.
+- `gzip64`: Gzip compression followed by Base64 encoding.
```
[the remark this answers is not in the document]
Comment (b35363) [reply 101.md:48]: Not directly, you don't have any compression by itself, you'll need Websocket Extension. Also, the only available compression algorithm is deflate, which is in some case not the best at all.

File: 101.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+- `gzip58`: Gzip compression followed by Base58 encoding.
```
[the remark this answers is not in the document]
Comment (b35363) [reply 101.md:49]: Like your comment, be creative, explain yourself. I stumbled upon cases where compression are a necessity.

**Comment by b35363** [comment]
> I don't get this, sounds like a terrible idea that we shouldn't do at all. You're compressing the JSON then base64ing it? In any case the same arguments I have against having a binary format and everything else also apply to this: it's not worth it.

I don't do this for you @fiatjaf, if you find it terrible just explain why, It's all about letting the choice. In some future case it might help you. 

People tend to burst up more and more data into nostr protocol, videos, long text, binary data... when you only deal with websockets of course it's not important, but by letting the choice, you could just make everyone happy. If you don't need to encode it, great don't use it, it's not mandatory. But when you'll need it, you might reconsider checking this NIP.

You really don't help the community by "governing" the nostr project

**Pull request closed without merging**